File: tasks-core-jvm/src/main/java/org/funfix/tasks/FiberExecutor.java
package org.funfix.tasks;

import org.jspecify.annotations.NullMarked;
import org.jspecify.annotations.Nullable;

import java.util.concurrent.Executor;
import java.util.concurrent.ThreadFactory;
import java.util.concurrent.atomic.AtomicReference;
import static org.funfix.tasks.ThreadPools.sharedIO;

@NullMarked
public interface FiberExecutor extends Executor {
    Cancellable executeCancellable(Runnable command, @Nullable Runnable onComplete);
    Fiber executeFiber(Runnable command);

    default Cancellable executeCancellable(Runnable command) {
        return executeCancellable(command, null);
    }

    static FiberExecutor fromThreadFactory(ThreadFactory factory) {
        return FiberExecutorDefault.fromThreadFactory(factory);
    }

    static FiberExecutor fromExecutor(Executor executor) {
        return FiberExecutorDefault.fromExecutor(executor);
    }

    static FiberExecutor shared() {
        return FiberExecutorDefault.shared();
    }
}

@NullMarked
@FunctionalInterface
interface RunnableExecuteFun {
    Cancellable invoke(Runnable command, @Nullable Runnable onComplete);
}

@NullMarked
final class FiberExecutorDefault implements FiberExecutor {
    private final Executor _executor;
    private final RunnableExecuteFun _executeFun;

    public FiberExecutorDefault(
            Executor executor,
            RunnableExecuteFun execute
    ) {
        _executor = executor;
        _executeFun = execute;
    }

    @Override
    public void execute(Runnable command) {
        _executor.execute(command);
    }

    @Override
    public Cancellable executeCancellable(Runnable command, @Nullable Runnable onComplete) {
        return _executeFun.invoke(command, onComplete);
    }

    @Override
    public Fiber executeFiber(Runnable command) {
        return SimpleFiber.create(_executeFun, command);
    }

    public static FiberExecutor fromThreadFactory(ThreadFactory factory) {
        final Executor executor = (command) -> {
            final var t = factory.newThread(command);
            t.start();
        };
        return new FiberExecutorDefault(
                executor,
                new RunnableExecuteFunViaThreadFactory(factory)
        );
    }

    public static FiberExecutor fromExecutor(Executor executor) {
        return new FiberExecutorDefault(
                executor,
                new RunnableExecuteFunViaExecutor(executor)
        );
    }

    @Nullable
    private static volatile FiberExecutor defaultFiberExecutorOlderJavaRef;
    @Nullable
    private static volatile FiberExecutor defaultFiberExecutorLoomRef;

    public static FiberExecutor shared() {
        if (VirtualThreads.areVirtualThreadsSupported()) {
            var ref = defaultFiberExecutorLoomRef;
            if (ref == null)
                synchronized (FiberExecutor.class) {
                    ref = defaultFiberExecutorLoomRef;
                    if (ref == null)
                        try {
                            ref = fromThreadFactory(VirtualThreads.factory());
                        } catch (VirtualThreads.NotSupportedException ignored) {
                        }
                    if (ref == null) {
                        ref = fromExecutor(sharedIO());
                    }
                    defaultFiberExecutorLoomRef = ref;
                }
            return ref;
        } else {
            var ref = defaultFiberExecutorOlderJavaRef;
            if (ref == null)
                synchronized (FiberExecutor.class) {
                    ref = defaultFiberExecutorOlderJavaRef;
                    if (ref == null) {
                        ref = fromExecutor(sharedIO());
                    }
                    defaultFiberExecutorOlderJavaRef = ref;
                }
            return ref;
        }
    }
}

@NullMarked
final class RunnableExecuteFunViaThreadFactory implements RunnableExecuteFun {
    private final ThreadFactory _factory;

    public RunnableExecuteFunViaThreadFactory(final ThreadFactory factory) {
        _factory = factory;
    }

    @Override
    public Cancellable invoke(Runnable command, @Nullable Runnable onComplete) {
        final var t = _factory.newThread(() -> {
            try {
                command.run();
            } finally {
                if (onComplete != null)
                    onComplete.run();
            }
        });
        t.start();
        return t::interrupt;
    }
}

@NullMarked
final class RunnableExecuteFunViaExecutor implements RunnableExecuteFun {
    private final Executor _executor;

    public RunnableExecuteFunViaExecutor(Executor executor) {
        _executor = executor;
    }

    @Override
    public Cancellable invoke(final Runnable command, final @Nullable Runnable onComplete) {
        final var state = new AtomicReference<State>(State.NotStarted.INSTANCE);
        _executor.execute(() -> {
            final var running = new State.Running(Thread.currentThread());
            if (!state.compareAndSet(State.NotStarted.INSTANCE, running)) {
                if (onComplete != null) onComplete.run();
                return;
            }
            try {
                command.run();
            } catch (final Exception e) {
                UncaughtExceptionHandler.logException(e);
            }

            while (true) {
                final var current = state.get();
                if (current instanceof State.Running) {
                    if (state.compareAndSet(current, State.Completed.INSTANCE)) {
                        if (onComplete != null) onComplete.run();
                        return;
                    }
                } else if (current instanceof State.Interrupting interrupting) {
                    try {
                        interrupting.wasInterrupted.await();
                    } catch (final InterruptedException ignored) {
                    } finally {
                        if (onComplete != null) onComplete.run();
                    }
                    return;
                } else {
                    throw new IllegalStateException("Invalid state: " + current);
                }
            }
        });
        return () -> triggerCancel(state);
    }

    private void triggerCancel(AtomicReference<State> state) {
        while (true) {
            final var current = state.get();
            if (current instanceof State.NotStarted) {
                if (state.compareAndSet(current, State.Completed.INSTANCE))
                    return;
            } else if (current instanceof State.Running running) {
                // Starts the interruption process
                final var wasInterrupted = new AwaitSignal();
                if (state.compareAndSet(current, new State.Interrupting(wasInterrupted))) {
                    running.thread.interrupt();
                    wasInterrupted.signal();
                    return;
                }
            } else {
                return;
            }
        }
    }

    sealed interface State {
        record NotStarted() implements State {
            public static NotStarted INSTANCE = new NotStarted();
        }
        record Completed() implements State {
            public static NotStarted INSTANCE = new NotStarted();
        }
        record Running(Thread thread) implements State {}
        record Interrupting(AwaitSignal wasInterrupted) implements State {}
    }
}

File: tasks-core/src/jvmMain/kotlin/org/funfix/tasks/FiberExecutor.kt
package org.funfix.tasks

import org.funfix.tasks.internals.*
import java.util.concurrent.Executor
import java.util.concurrent.ThreadFactory
import java.util.concurrent.atomic.AtomicReference

interface FiberExecutor: Executor, ExecuteCancellableFun {
    fun executeFiber(command: Runnable): Fiber

    companion object {
        @JvmStatic
        fun fromThreadFactory(factory: ThreadFactory): FiberExecutor =
            FiberExecutorDefault.fromThreadFactory(factory)

        @JvmStatic
        fun fromExecutor(executor: Executor): FiberExecutor =
            FiberExecutorDefault.fromExecutor(executor)

        @JvmStatic
        fun shared(): FiberExecutor =
            FiberExecutorDefault.shared()
    }
}

fun interface ExecuteCancellableFun {
    fun executeCancellable(command: Runnable, onComplete: Runnable?): Cancellable
}

private class FiberExecutorDefault private constructor(
    private val _executeFun: ExecuteCancellableFun,
    private val _executor: Executor
) : FiberExecutor {
    override fun executeFiber(command: Runnable): Fiber {
        val fiber = ExecutedFiber()
        val token = executeCancellable(command) { fiber.signalComplete() }
        fiber.registerCancel(token)
        return fiber
    }

    override fun executeCancellable(command: Runnable, onComplete: Runnable?): Cancellable =
        _executeFun.executeCancellable(command, onComplete)

    override fun execute(command: Runnable) =
        _executor.execute(command)

    companion object {
        fun fromThreadFactory(factory: ThreadFactory): FiberExecutor =
            FiberExecutorDefault(
                ExecuteCancellableFunViaThreadFactory(factory)
            ) { command ->
                val t = factory.newThread(command)
                t.start()
            }

        fun fromExecutor(executor: Executor): FiberExecutor =
            FiberExecutorDefault(
                ExecuteCancellableFunViaExecutor(executor),
                executor
            )

        @Volatile
        private var defaultFiberExecutorOlderJavaRef: FiberExecutor? = null
        @Volatile
        private var defaultFiberExecutorLoomRef: FiberExecutor? = null

        fun shared(): FiberExecutor {
            if (VirtualThreads.areVirtualThreadsSupported()) {
                // Double-checked locking
                if (defaultFiberExecutorLoomRef != null) return defaultFiberExecutorLoomRef!!
                synchronized(this) {
                    if (defaultFiberExecutorLoomRef != null) return defaultFiberExecutorLoomRef!!
                    try {
                        defaultFiberExecutorLoomRef = fromThreadFactory(VirtualThreads.factory())
                        return defaultFiberExecutorLoomRef!!
                    } catch (ignored: VirtualThreads.NotSupportedException) {}
                }
            }
            // Double-checked locking
            if (defaultFiberExecutorOlderJavaRef != null) return defaultFiberExecutorOlderJavaRef!!
            synchronized(this) {
                if (defaultFiberExecutorOlderJavaRef != null) return defaultFiberExecutorOlderJavaRef!!
                defaultFiberExecutorOlderJavaRef = fromExecutor(ThreadPools.sharedIO())
                return defaultFiberExecutorOlderJavaRef!!
            }
        }
    }
}

private class ExecuteCancellableFunViaThreadFactory(
    private val _factory: ThreadFactory
) : ExecuteCancellableFun {
    override fun executeCancellable(command: Runnable, onComplete: Runnable?): Cancellable {
        val t = _factory.newThread {
            try {
                command.run()
            } finally {
                onComplete?.run()
            }
        }
        t.start()
        return Cancellable { t.interrupt() }
    }
}

private class ExecuteCancellableFunViaExecutor(
    private val _executor: Executor
) : ExecuteCancellableFun {
    override fun executeCancellable(command: Runnable, onComplete: Runnable?): Cancellable {
        val state = AtomicReference<State>(State.NotStarted)
        _executor.execute {
            val running = State.Running(Thread.currentThread())
            if (!state.compareAndSet(State.NotStarted, running)) {
                onComplete?.run()
                return@execute
            }
            try {
                command.run()
            } catch (e: Exception) {
                UncaughtExceptionHandler.logOrRethrow(e)
            }

            while (true) {
                when (val current = state.get()) {
                    is State.Running ->
                        if (state.compareAndSet(current, State.Completed)) {
                            onComplete?.run()
                            return@execute
                        }
                    is State.Interrupting -> {
                        if (state.compareAndSet(current, State.Completed)) {
                            try {
                                current.wasInterrupted.await()
                            } catch (ignored: InterruptedException) {
                            } finally {
                                onComplete?.run()
                            }
                            return@execute
                        }
                    }
                    State.Completed, State.NotStarted ->
                        throw IllegalStateException("Invalid state: $current")
                }
            }
        }
        return Cancellable { triggerCancel(state) }
    }

    private fun triggerCancel(state: AtomicReference<State>) {
        while (true) {
            when (val current = state.get()) {
                is State.NotStarted ->
                    if (state.compareAndSet(current, State.Completed)) {
                        return
                    }
                is State.Running -> {
                    val wasInterrupted = AwaitSignal()
                    if (state.compareAndSet(current, State.Interrupting(wasInterrupted))) {
                        current.thread.interrupt()
                        wasInterrupted.signal()
                        return
                    }
                }
                is State.Completed ->
                    return
                is State.Interrupting ->
                    throw IllegalStateException("Invalid state: $current")
            }
        }
    }

    sealed interface State {
        data object NotStarted : State
        data class Running(val thread: Thread) : State
        data class Interrupting(val wasInterrupted: AwaitSignal) : State
        data object Completed : State
    }
}
